### ISSUE TITLE
Batch updating cells can exceed quota for reading

### ISSUE THREAD

**Issue by Kirkman** [burnash/gspread#609]
**GitHub Issues are for bugs and feature requests.**

### Environment info

Operating System: Mac OS X 10.11.6
Python version: 2.7.14
gspread version: 3.0.1

### Steps to reproduce
1. Use code similar to what is shown in the Docs under [Updating Cells](https://gspread.readthedocs.io/en/latest/user-guide.html#updating-cells)

Within a loop where I'm looking for cells to update, add any to a list:

```
if needs_update == True:
	updated_cell = wks.cell( row_num, col_num )
	updated_cell.value = new_valuer
	cells_to_update.append( updated_cell )
```

At the end of the script, I will send the list to be batch updated:
```
wks.update_cells( cells_to_update )
```

2. The problem occurs BEFORE I get to the `update_cells()` call. I get a "code 429" and "Quota exceeded for quota group 'ReadGroup' and limit 'USER-100s' of service 'sheets.googleapis.com'"

3. I believe what's happening is that every time I try to fetch a cell instance by doing `updated_cell = wks.cell( row_num, col_num )`, the `.cell()` method runs `self.spreadsheet.values_get()`, which results in an API call. 

So, if I'm understanding this correctly, for every cell I want to update, I will have to make one API call to read its value. Hundreds of cells to update? Hundreds of API calls to read. All these API calls at once causes me to exceed the "ReadGroup" quota limit.

Is there a way to suppress this? A way to get a Cell instance without an API call? Or perhaps a way to batch read a list of cells in one API call?

**Comment by te-evalenzuela** [comment]
Is there a reason you can't pull the entire sheet(1 call), loop through to create a list of cell updates needed, and send a batch call to update(1 call)?

**Comment by Kirkman** [comment]
Maybe?

I just think it's weird that the recommended code in the docs for doing a batch update does indeed allow for a single batch _write_, but can potentially have hundreds of individual _reads_.

**Comment by burnash** [comment]
Could you please show the example you're referring to? The code in [Updating Cells](https://gspread.readthedocs.io/en/latest/user-guide.html#updating-cells) at the moment looks like this:

```python
# Select a range
cell_list = worksheet.range('A1:C7')

for cell in cell_list:
    cell.value = 'O_o'

# Update in batch
worksheet.update_cells(cell_list)
```

This will perform one API call in `worksheet.range` and one call in `worksheet.update_cells`.
Let me know if I'm missing something.

**Comment by burnash** [comment]
I'm closing this issue due to inactivity. Feel free to reopen it if you have an update.

**Comment by Kirkman** [comment]
Sorry for the delay, @burnash. 

To answer your question, yes, that is the example I'm talking about. The difference is that I want to batch_update a list of _non-contiguous_ cells. 

As far as I can tell, gspread's `range()` will not allow me to pass a list of non-contiguous cells. It only allows rectangular selections.

So, while I can do a single WRITE via `wks.update_cells()`, I would have to read the cells individually using `wks.cell()`, resulting in hundreds of API calls to read the cell values.

Things I'm wondering about, which you can feel free to dismiss:

* Could `range()`  be updated to allow non-rectangular selections, such as a list of individual cells? (Maybe via [rangeList](https://developers.google.com/apps-script/reference/spreadsheet/range-list) or [batchGet](https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets.values/batchGet) ?)
* Why doesn't `range()` support columnar selections, like `range('A:B')`? In my use case, this would help since the number of rows in the sheet can change dynamically.
* Is there a way to either fetch or create a Cell() instance *without* an API call? For example, if I'm just overwriting cells, I really don't care about the old value. I just want to write the new value. 

FWIW, I tried @te-evalenzuela's suggestion of pulling the entire sheet and looping over the cells. I tried doing this using worksheet's `._fetch_cells()` method, but it doesn't retrieve blank cells! 

I have been able to get this method to work by doing something like this:
```
	feed = wks.client.get_cells_feed( worksheet=wks, params={'return-empty':'true'} )
	cells = [Cell(wks, elem) for elem in feed.findall(_ns('entry'))]
```

It just seems really goofy that I should have to do all that when I know up front which cells I want to fetch.

**Comment by Kirkman** [comment]
Okay, I sort of monkey-patched some code together that will do a batchUpdate of non-contiguous cells. It basically takes the list of cells and turns it into a list of value_ranges, then sends the whole thing to the batchUpdate endpoint.

I'm sure this is not the optimal way to do this. My hope is that it might stir you to cook up something better for gspread.


```
def values_batch_update(self, body=None):

	SPREADSHEET_VALUES_BATCH_URL = 'https://sheets.googleapis.com/v4/spreadsheets/%s/values:batchUpdate'

	url = SPREADSHEET_VALUES_BATCH_URL % (self.id)

	r = self.client.request('post', url, json=body)
	return r.json()

# Add this method to the Spreadsheet class
gspread.models.Spreadsheet.values_batch_update = values_batch_update


def update_cells_noncontiguous(self, cell_list, value_input_option='RAW'):
	value_ranges = []
	for cell in cell_list:
		value_range = {
			"range": self.title + '!' + gspread.utils.rowcol_to_a1( cell.row, cell.col ),
			"values": [
				[ cell.value ]
			],
		}
		value_ranges.append( value_range )

	data = self.spreadsheet.values_batch_update(
		body={
			"valueInputOption": value_input_option,
			"data": value_ranges,
		},
	)

	return data

# Add this method to the Worksheet class
gspread.models.Worksheet.update_cells_noncontiguous = update_cells_noncontiguous

```

**Comment by ndevenish** [comment]
I think I've got this same need, also ran into read quota throttles when repeatedly calling `.get_cell`. I then was going to try building a list of `Cell` objects to pass to update cells but thanks to this thread know that apparently behaves unexpectedly.

I've avoided repeated reads by reading the whole sheet at once, but want to update an effectively random sample of cells that may or may not be partially contiguous - I'd also like to avoid just bulk updating the whole sheet for versioning if possible, unless that's the right way to approach this.

@Kirkman's modification seems to work for what I need; one call updating lots of disparate things. Is this a good way to approach the problem or is there a better way to do it?

Should we be putting together a pull request for this?

**Comment by taewookim** [comment]
Having same problem here

**Comment by taewookim** [comment]
temporary solution that seems to work (for now). add a sleep statement:

```
for cell in cell_list:
    cell.value = 'O_o'
    time.sleep(1) # or whatever you see fit

# Update in batch
worksheet.update_cells(cell_list)
```

**Comment by akshanshshrivastava** [comment]
I am having the same issue.  I want to change the background color of the selected cell_list, which is more than 200 in my code:
I am trying using try-except to check If I exceed the limit I will log in again and try again. but this is then skipping some of the cells to get formatted

```
import time
from oauth2client.service_account import ServiceAccountCredentials
import gspread
from gspread_formatting import CellFormat, Color, TextFormat, format_cell_range

yellow_fmt = CellFormat(backgroundColor=Color(1, 1, 0), textFormat=TextFormat(
    bold=True, foregroundColor=Color(0, 0, 0)), horizontalAlignment="CENTER")

scope = ['https://spreadsheets.google.com/feeds',
         'https://www.googleapis.com/auth/drive']

credentials = ServiceAccountCredentials.from_json_keyfile_name(
    'credentialfile.json', scope)
gc = gspread.authorize(credentials)
spreadsheetObject = gc.open_by_key("my spreadsheet id")
worksheet = spreadsheetObject.worksheet("wks")

three = worksheet.findall("3")
cell_list = []
for cell in three:
    celladdress = gspread.utils.rowcol_to_a1(cell.row, cell.col)
    cell_list.append(celladdress)
while True:
    try:
        for cell in cell_list:
            print(cell)
            format_cell_range(worksheet, cell, yellow_fmt)
    except gspread.exceptions.APIError:
        time.sleep(101)
        gc.login()
        continue
    break
```

**Comment by akshanshshrivastava** [comment]
> temporary solution that seems to work (for now). add a sleep statement:
> 
> ```
> for cell in cell_list:
>     cell.value = 'O_o'
>     time.sleep(1) # or whatever you see fit
> 
> # Update in batch
> worksheet.update_cells(cell_list)
> ```

it will work but it is time consuming there should be me a method to batch update all the cells at once with some condition.

**Comment by taewookim** [comment]
@burnash Should this be closed w/ release of v3.3.0?

**Comment by burnash** [comment]
@taewookim correct,

Since #731 / [v3.3.0](https://github.com/burnash/gspread/blob/master/HISTORY.rst#330-2020-03-12) there are new methods for batch fetching or batch updating cell values:

*Worksheet.batch_get()*

Example:

```python
worksheet.batch_get(['A1:B2', 'F12'])
```
– returns list of `ValueRange` objects (which are in turn lists of lists).

*Worksheet.batch_update()*

Example:

```python
worksheet.batch_update([{
    'range': 'A1:B1',
    'values': [['42', '43']],
}, {
    'range': 'my_range',
    'values': [['44', '45']],
}])
```

In both examples `worksheet` is an instance of the `Worksheet` class.

More examples will follow in the [docs](https://gspread.readthedocs.io/en/latest/).
